GeneralsMD/Code/GameEngine/Source/Common/Thing/Thing.cpp

```diff
@@ -208,40 +208,26 @@ void Thing::setOrientation( Real angle )
 	Coord3D oldPos = m_cachedPos;
 	Matrix3D oldMtx = m_transform;
 
-	pos.x = m_transform.Get_X_Translation();
-	pos.y = m_transform.Get_Y_Translation();
-	pos.z = m_transform.Get_Z_Translation();
-	if( m_template->isKindOf( KINDOF_STICK_TO_TERRAIN_SLOPE) )
+	m_cachedAngle = normalizeAngle(angle);
+	Coord3D pos = { m_transform[0][3], m_transform[1][3] ,m_transform[2][3] };
+	m_cachedPos = pos;
+
+	if (m_template->isKindOf(KINDOF_STICK_TO_TERRAIN_SLOPE))
 	{
-		Matrix3D mtx;
-		const Bool stickToGround = true;	// yes, set the "z" pos				
-		TheTerrainLogic->alignOnTerrain(angle, pos, stickToGround, m_transform );
+		TheTerrainLogic->alignOnTerrain(angle, m_cachedPos, true, m_transform);
 	}
 	else
 	{
-		z.x = 0.0f;
-		z.y = 0.0f;
-		z.z = 1.0f;
-
-		u.x = Cos(angle);
-		u.y = Sin(angle);
-		u.z = 0.0f;
-
-		y.crossProduct( &z, &u, &y );
-		x.crossProduct( &y, &z, &x );
-
-		m_transform.Set(  x.x, y.x, z.x, pos.x,
-											x.y, y.y, z.y, pos.y,
-											x.z, y.z, z.z, pos.z );
+		Real a = Sin(angle);
+		Real b = Cos(angle);
+		m_transform[0][0] = b; m_transform[0][1] = -a; m_transform[0][2] = 0;
```

Comment:
Why?
if you thought this is cheaper, nope, m_transform[?][?] actually makes extra calls, Matrix3D::operator[] Vector4::operator[]